File: src/org/elixir_lang/debugger/node/TraceElement.java
package org.elixir_lang.debugger.node;

import com.ericsson.otp.erlang.OtpErlangList;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;

public class TraceElement {
  private final String myModule;
  private final String myFunction;
  private final OtpErlangList myFunctionArgs;
  private final Collection<VariableBinding> myBindings;
  private final String myFile;
  private final int myLine;

  public TraceElement(@NotNull String module, @NotNull String function, @NotNull OtpErlangList functionArgs,
                      @NotNull Collection<VariableBinding> bindings, String file, int line) {
    myModule = module;
    myFunction = function;
    myFunctionArgs = functionArgs;
    myBindings = bindings;
    myFile = file;
    myLine = line;
  }

  @NotNull
  public String getModule() {
    return myModule;
  }

  @NotNull
  public String getFunction() {
    return myFunction;
  }

  @NotNull
  public OtpErlangList getFunctionArgs() {
    return myFunctionArgs;
  }

  @NotNull
  public Collection<VariableBinding> getBindings() {
    return myBindings;
  }

  public String getFile() {
    return myFile;
  }

  public int getLine() {
    return myLine;
  }
}

File: src/org/elixir_lang/debugger/node/TraceElement.kt
package org.elixir_lang.debugger.node

import com.ericsson.otp.erlang.OtpErlangObject
import com.ericsson.otp.erlang.OtpErlangTuple
import com.intellij.openapi.diagnostic.Logger
import org.elixir_lang.beam.term.inspect

class TraceElement(
        val level: Int,
        val module: String,
        val function: String,
        val arguments: List<OtpErlangObject>,
        val bindings: List<Binding>,
        val file: String,
        val line: Int
) {
    companion object {
        // {level, {module, function, arguments}, bindings, {file, line}}
        private const val ARITY = 4
        private val LOGGER = Logger.getInstance(TraceElement::class.java)

        fun from(otpStackFrame: OtpErlangObject): TraceElement? =
            when (otpStackFrame) {
                is OtpErlangTuple -> from(otpStackFrame)
                else -> {
                    LOGGER.error("OTP Stack Frame (${inspect(otpStackFrame)}) is not an OtpErlangTuple")

                    null
                }
            }

        private fun from(otpStackFrame: OtpErlangTuple): TraceElement? {
            val arity = otpStackFrame.arity()

            return if (arity == ARITY) {
                Level.from(otpStackFrame.elementAt(0))?.let { level ->
                    ModuleFunctionArguments.from(otpStackFrame.elementAt(1))?.let { (module, function, arguments) ->
                        Bindings.from(otpStackFrame.elementAt(2))?.let { bindings ->
                            FileLine.from(otpStackFrame.elementAt(3))?.let { (file, line) ->
                                TraceElement(level, module, function, arguments, bindings, file, line)
                            }
                        }
                    }
                }
            } else {
                LOGGER.error("OTP Stack Frame (${inspect(otpStackFrame)}) arity ($arity) is not expected ($ARITY)")

                null
            }
        }
    }
}
